File: 210. Course Schedule II.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(numCourses, prerequisites) {
    var coursePriority = [];
    var courseSubsequent = [];

    for ( var i = 0; i < numCourses; i ++ ) {
        coursePriority[i] = 0;
        courseSubsequent[i] = [];
    }

    for ( var j = 0; j < prerequisites.length; j++ ) {
        coursePriority[prerequisites[j][0]] ++;
        courseSubsequent[prerequisites[j][1]].push(prerequisites[j][0]);
    }

    var queue = [];
    for ( var k = 0; k < coursePriority.length; k++ ) {
        if ( coursePriority[k] == 0 ) {
            queue.push(k);
        }
    }

    var result = [];
    while ( queue.length > 0 ) {
        var course = queue.shift();
        result.push( course );
        var n = courseSubsequent[course].length;
        if ( n > 0 ) {
            for ( var l = 0; l < n; l ++ ) {
                var pointer = courseSubsequent[course][l];
                coursePriority[pointer]--;
                if ( coursePriority[pointer] == 0 ) {
                    queue.push(pointer);
                }
            }
        }
    }

    if ( result.length == numCourses ) {
        return result
    } else {
        return [] ;
    }
};

File: 210. Course Schedule II.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        p = [0] * numCourses
        r = [[] for _ in xrange(numCourses)]
        for prereq in prerequisites:
            p[prereq[0]] += 1
            r[prereq[1]].append(prereq[0])
        q = []
        for i in xrange(len(p)):
            if p[i] == 0:
                q.append(i)
        result = []
        while q:
            n = q.pop(0)
            result.append(n)
            for child in r[n]:
                p[child] -= 1
                if p[child] == 0:
                    q.append(child)
        return result if len(result) == numCourses else []
